content/general/concepts/programming-optimization/programming-optimization.md

```diff
@@ -0,0 +1,265 @@
+
+Also important key points to keep in mind before optimizing the source code: 
+- Time based optimization can execute faster output, but it can also lead to a bigger code base, which can create conflict with memory and consumption. Consideration on the requirements needed will play an important role on this one.
+- Performance optimization is a never-ending process. There will always be improvement opportunities for faster source code, especially as technology keeps moving ahead. 
+- Let's not forget that temptation to use certain optimization methods in programming to execute faster outputs and can result in best coding standards and practices not be followed. Avoiding such methods will render better results long term..
+
+Shall we look at some examples of different types of source coding optimizations and their modest overview? 
+
+Yes? Ok, let's do it!
+
+**Loop Optimization**: majority of system programs, if not all operate in loops. It is vital to optimize these to save CPU cycles and memory, avoiding loop-invariants and performing induction analysis will have a positive impact on optimization.
+
+Example:
+
+```javascript
+do { i = 10;  v = v + i; } 
+while (v<100);
+```
+
+The code above repeats the assignment of the identifier i for every iteration, which if we put it this way:
+
+```javascript
+i = 10;
+do { v = v + i; } 
+while (v<100);
+```
+
+this should not only save the CPU cycles but can also be used on any processor.
+
+**Strength Reduction**: Some expressions use more CPU cycles, time, and memory than others. These expressions should be replaced with less expensive expressions without sacrificing the output(e.g. multiplication (x * 2) uses more CPU cycles than (x 1) but producing the same output).
+
+**Partially Dead Code**
+This type of source code refers to statements that include value calculations that are used only in particular and irregular conditions.
+
+**Unreachable Code Elimination**
+Before implementing this type of *source code optimization*, ideally a control flow graph should be created first. Only after determining that a particular inaccessible code block does not have an incoming edge, these branches can then be deleted after continual propagation and folding.
+
+**Function Inlining** 
+The body of the function takes the place of a function call. This saves a lot of time by eliminating the need to copy all parameters, store the return address, and so on.
+
+**Function Cloning** 
+For different calling arguments, specialized source codes functions are implemented(e.g. overloading a function is an example of this).
+
+**Partial Redundancy**
+In a parallel route, redundant expressions are calculated many times without changing the operands and partial-redundant, on the other hand, are calculated several times along a path without changing the operands. A code-motion approach, loop-invariant can possibly result in a largely redundant render.
+
+Beverage anyone? Cup of tea for me please...
+
+
+### Builds
```

Comment:
```suggestion
### Builds

```

Reply:
You still need to make this edit. There needs to be a line break before and after each header.